README.md
# What is this?

**It is a nice coding test that I've taken recently.**

> "Zombie apocalypse is upon us. You can't take a ride safely anymore."

This is a sets of cloud-based microservices that consume and persist geolocated data sent by a mobile device to discover if the owner (a taxi driver) is a zombie or not. 

> A driver is a zombie if he has driven less than 500 meters in the last 5 minutes.

## What are those services? 
### 1. Gateway Service

The `Gateway` service is a _public facing service_.
HTTP requests hitting this service are either transformed into [NSQ](https://github.com/nsqio/nsq) messages or forwarded via HTTP to specific services.

The service must be configurable dynamically by loading the provided `gateway/config.yaml` file to register endpoints during its initialization.

#### Public Endpoints

`PATCH /drivers/:id/locations`

**Payload**

```json
{
  "latitude": 48.864193,
  "longitude": 2.350498
}
```

**Role:**

During a typical day, thousands of drivers send their coordinates every 5 seconds to this endpoint.

**Behaviour**

Coordinates received on this endpoint are converted to [NSQ](https://github.com/nsqio/nsq) messages listened by the `Driver Location` service.

---

`GET /drivers/:id`

**Response**

```json
{
  "id": 42,
  "zombie": true
}
```

**Role:**

Users request this endpoint to know if a driver is a zombie.
A driver is a zombie if he has driven less than 500 meters in the last 5 minutes.

**Behaviour**

This endpoint forwards the HTTP request to the `Zombie Driver` service.

### 2. Driver Location Service
The `Driver Location` service is a microservice that consumes drivers' location messages published by the `Gateway` service and stores them in a Redis database.

It also provides an internal endpoint that allows other services to retrieve the drivers' locations, filtered and sorted by their addition date

#### Internal Endpoint

`GET /drivers/:id/locations?minutes=5`

**Response**

```json
[
  {
    "latitude": 48.864193,
    "longitude": 2.350498,
    "updated_at": "2018-04-05T22:36:16Z"
  },
  {
    "latitude": 48.863921,
    "longitude":  2.349211,
    "updated_at": "2018-04-05T22:36:21Z"
  }
]
```

**Role:**

This endpoint is called by the `Zombie Driver` service.

**Behaviour**

For a given driver, returns all the locations from the last 5 minutes (given `minutes=5`).


### 3. Zombie Driver Service
The `Zombie Driver` service is a microservice that determines if a driver is a zombie or not according to the previously stated definition.

#### Internal Endpoint

`GET /drivers/:id`

**Response**

```
{
  "id": 42,
  "zombie": true
}
```

**Role:**

This endpoint is called by the `Gateway` service.

**Behaviour**

Returns the zombie state of a given driver. 


# Setting up 
## Premises  
### Environment:
The following assumptions has been made for testing environment
- The Services use the "0" database in Redis 
- AUTH is disabled (no password set)

## How to install it

### 1. Clone repository
1) `git clone git@github.com:silvestriluca/zombie-drivers.git`

2) `cd zombie-drivers `

### 2. Starts the NSQ and REDIS services
1) Write down the IP of a machine running docker (e.g. if you use `docker-machine` issue the command `docker-machine ip`)

2) Open the file  `docker-compose.yml`

3) Edit the following entry under `nsqd` service: `command: /nsqd --lookupd-tcp-address=nsqlookupd:4160 --broadcast-address=YOUR-DOCKER-MACHINE-IP-GOES-HERE`

4) Save `docker-compose.yml`

5) In a terminal, navigate to repo root directory and issue the following command to launch Redis and NSQ: `docker-compose up`

6) Open a browser and navigate to `http://YOUR-DOCKER-MACHINE-IP-GOES-HERE:4171/`. Nsqadmin dashboard should launch

7) Check the `NODES` tab. If everything went good, you should see a `Hostname` with `Broadcast address` pointing to `YOUR-DOCKER-MACHINE-IP-GOES-HERE` 
### 3. Build Gateway
From the **root repository directory**:

1) `cd gateway`

2) `go get`  => installs the required packages

3) `go build`  => The executable is built and created in the current directory

4) Edit `config.yaml` file to set the gateway parameters. Choose a port that is not already occupied (e.g. `3000`)
### 4. Build Driver-location
From the **root repository directory**:

1) `cd driver-location`

2) `go get` => installs the required packages

3) `go build` => The executable is built and created in the current directory

4) Edit `config.yaml` file to set the service parameters. Choose a different port than `Gateway service` (e.g. `3001`)

### 5. Build Zombie-driver
From the **root repository directory**:

1) `cd zombie-driver`

2) `go get` => installs the required packages

3) `go build` => The executable is built and created in the current directory

4) Edit `config.yaml` file to set the service parameters. Choose a different port than `Gateway service` and  `Driver location service` (e.g. `3002`)

5) To **dynamically set** "Zombie definitions" save the following keys in Redis "0" database
- `SET zombie-e [value]`  => Timespan (in minutes) to evaluate a zombie state (default = 5 min)

- `SET zombie-mdc [value]`  => Maximum distance (in meters) that a zombie can cover zombie-e timespan (default = 500 m)

## How to launch the services
1) If you've followed the **HOW TO INSTALL** section, you should have 3 executables in 
- `REPOSITORY_DIR/gateway` 
- `REPOSITORY_DIR/driver-location` 
- `REPOSITORY_DIR/zombie-driver`  

2) Be sure that the **executables have exec permissions**.

3) Be sure that REDIS and NSQ services are running. If not, from the REPOSITORY_DIR root directory issue `docker-compose start`

4) Be sure that Redis has no password set (no AUTH)

5) To launch a service, from REPOSITORY_DIR root directory:

 - `cd SERVICENAME`
 - `SERVICENAME`

## Tests
As of now, tests don't mock external services, so they REQUIRE all the services running (gateway, driver-location, zombie-driver, redis, nsq).
Tests can be executed by running in the service directory the command `go test`

Example: From the root REPOSITORY_DIR
- `cd gateway` 
- `go test` 

# Description 
### Environment assumptions (reasonable for a testing environment):
The following assumptions have been made during development and they can be easily removed with little code changes.
- The Demo use the "0" database in Redis 
- AUTH is disabled (no password set. AUTH isn't implemented)
- All the communications happen in the backend and are unencrypted (no TLS). For the Gateway that means that there is a webserver in front of it, like Apache or Nginx configured like a proxy with proper certificates to ensure https on public-facing Rest APIs.
See also Gateway description.

### Gateway
The gateway has endpoints and exposed port fully configurable. Configuration happens by modifying the `config.yaml` file, located in the same directory as the executable. The Config object (struct) holds all the configurations and can be used in the future for different config source (e.g. command line aguments using flag package). At this moment, the path+name of the config file can be changed only by editing the global const `ConfigFileName`.

To manage a scalable routing system the choice felt on GIN (<https://gin-gonic.github.io/gin/>) as it is listed as one of the highest performance router out there. Every endpoint is managed by a concurrent handler.

Gateway is http-only in its current setup . Having a front proxy (Apahe/Nginx) configured with server certificates is a common setup, that is why I've implemented an http-only gateway. Changing this design choice is very easy, since Gin can be quickly configured to serve the endpoints on https.

YAML parsing has been implemented by adopting the widely used <https://gopkg.in/yaml.v2>

Posting to NSQ topic is made using a HTTP POST request to a nsqd http `/pub` endpoint. The call happens in the handler called by GIN route (by default `/drivers/:id/locations`). 
Using a handler ensures scalability (GIN takes care to spawn an underlying goroutine at every handler call).
A possible improvement is using the nsq client and its Producer object, which would guarantee a better topology abstraction by using nsqdlookup directly to discover nsqd servers.

Since the `/drivers/:id/locations` call saves the location informations asyncronously, a preliminary validation on payload (body) is being made and a 400 error is returned if the payload is malformed or lacks informations.

**TESTS** cover most of the code but they rely on working services. An improvement to them could be a full mocking of the other services, but this hasn't been implemented in this initial commit.

All the codebase (service and tests) is fully commented to be easily readable and self-explaining.

### Driver Location Service
The driver-location service is configurable through a `config.yaml` file, located in the same directory as the executable. Settings include listening port, nsq and redis related informations. The Config object (struct) holds all the configurations and can be used in the future for different config source (e.g. command line aguments using flag package). At this moment, the path+name of the config file can be changed only by editing the global const `ConfigFileName`.

To manage a scalable routing system the choice felt on GIN (<https://gin-gonic.github.io/gin/>) for the same reasons stated for the gateway service. Every endpoint is managed by a concurrent handler.

Service endpoints are http-only as it is supposed to run in an isolated backend behind a proxy and gateway. As for the gateway, changing this design choice is very easy, since Gin can be quickly configured to serve the endpoints on https.

YAML parsing has been implemented by adopting the widely used <https://gopkg.in/yaml.v2>

NSQ interaction happens by implementing a Consumer object, as provided by the official nsq go client (<https://github.com/nsqio/go-nsq>). The client pools nsqlookupd to discover nsqds that provides the specified topic.

A (concurrent) handler is attached to receive message event. The number of concurrent publishers can be set in config file (`num-publishers`). Scalability is assured by the concurrent handler approach. Some load tests and number of driver estimations should be done to define the ideal `num-publishers` value, taking in consideration also more driver-location-service instances.

Redis interaction is done by using the well-known Redigo client (<https://github.com/gomodule/redigo>).
A connection pool is initialized at service launch. When a redis related operation is needed, a connections is taken from the pool and sent back to it when operation ends. 

With the code settings there are no limits on the number of connections in the pool while the maximum number of idle connections in the pool is 3, with an idle timeout = 240s

The service makes use of Redis geospatial features (GEOADD)

The Redis data structure is described in the following [section](#data). There it is described also how the service retrieves the locations requested with `GET /drivers/:id/locations?minutes=5`

A validator function takes care of parsing the message input for errors/problems before persisting it to the database.

Requests to the `GET /drivers/:id/locations?minutes=5` endpoint relative to non-existing drivers result in HTTP 404 errors.

Internal errors result in HTTP 500 response.

Errors are managed (as far as I've tested).

For **TESTS**, same considerations as described for Gateway are valid.

All the codebase (service and tests) is fully commented to be easily readable and self-explaining.

**BONUS:** An optional `distance=true` querystring option has been integrated for endpoint `GET /drivers/:id/locations`. Internally it uses GEODIST (as described in [this section](#data)) and allows to recover the elapsedDistance and cumulativeDistance in the last Z minutes timespan. For example:

`GET /drivers/:id/locations?distance=true&minutes=5` 
```
[
    {
        "cumulativeDistance": 0,
        "elapsedDistance": 0,
        "latitude": 48.864193,
        "longitude": 2.364986,
        "updated_at": "2018-10-24T13:58:10Z"
    },
    {
        "cumulativeDistance": 73.4,
        "elapsedDistance": 73.4,
        "latitude": 48.864193,
        "longitude": 2.365989,
        "updated_at": "2018-10-24T13:58:24Z"
    },
    {
        "cumulativeDistance": 146.407,
        "elapsedDistance": 73.007,
        "latitude": 48.864193,
        "longitude": 2.366987,
        "updated_at": "2018-10-24T14:00:17Z"
    }
]
```

### Zombie Driver Service
The zombie-driver service is configurable through a `config.yaml` file, located in the same directory as the executable. Settings include listening port, Redis and Driver-Location-Service related informations. The Config object (struct) holds all the configurations and can be used in the future for different config source (e.g. command line aguments using flag package). At this moment, the path+name of the config file can be changed only by editing the global const `ConfigFileName`.

To manage a scalable routing system the choice felt on GIN (<https://gin-gonic.github.io/gin/>) for the same reasons stated for the gateway service. Every endpoint is managed by a concurrent handler.

Service endpoints are http-only as it is supposed to run in an isolated backend behind a proxy and gateway. As for the gateway, changing this design choice is very easy, since Gin can be quickly configured to serve the endpoints on https.

YAML parsing has been implemented by adopting the widely used <https://gopkg.in/yaml.v2>

The service uses Redigo client (<https://github.com/gomodule/redigo>). Same pool approach than Driver Location Service has been followed.

To evaluate "zombie state", a distance has to be computed. By default it tries to retrive it from Driver-Location Service using `distance=true` querystring option. If the answer doesn't include distance informations, the zombie-driver service computes it using GEODIST Redis function, as described in [this section](#data).

Requests to the `GET /drivers/:id` endpoint relative to non-existing drivers result in HTTP 404 errors.

Internal errors / missing-upstream-service errors result in HTTP 5xx response (500/503).

Errors are managed (as far as I tested).

For **TESTS**, same considerations as described for Gateway are valid.

All the codebase (service and tests) is fully commented to be easily readable and self-explaining.

**BONUS:** "Zombie parameters" (max distance D covered in time Z) default as per initial service description (D=500m , Z=5s) but can be optionally set in Redis.

The zombie definition is configurable on fly through 2 REDIS key-values:
- **zombie-e** => Timespan (in minutes) to evaluate a zombie state (default = 5 min)
- **zombie-mdc** =>  Maximum distance (in meters) that a zombie can cover during zombie-e timespan (default = 500 m) 

## Redis: Driver related data structure and how it is used by the services<a name="data"></a>
Everytime a driver sends his/her location, the following keys are populated:
1) `on-course` => GEOADD longitude, latitude, **driverId**
2) `driver:<driverId>:log` => GEOADD longitude, latitude, **UnixTimestamp**
3) `driver:<driverId>:timestamps` => SADD **UnixTimestamp**

(1) and (2) are geohashes (sorted sets with special methods on it, like GEODIST).

(3) is a set

- (1) stores the last known position of every driver
- (2) stores the position of driverId at a given timestamps
- (3) stores all the recorded timestamps of a given driverId

Recovering  the position lists in the last Z minutes is as easy as:

1) retrieve from (3) the timestamps list with a SORT commmand ordered by DESC (from the newest to the oldest). Limit the search to maximum of Z*60 elements (since Unix timestamp resolution = 1 sec)
2) reduce the list to include only timestamps inside the required timespan
3) Use GEOPOS command on (2) to every timestamp entry in the reduced list 

Evaluating the distance covered by a driverId in a given timespan it's even easier: use the same procedure described before to retrieve relevant timestamps and use GEODIST command to evaluate *delta* distance between two consequent timestamps in the reduced list, iterating on timestamps and cumulating the *deltas*. 

## BONUSES (optional features) :confetti_ball:
### Bonus point 1
The zombie definition is configurable on fly through 2 REDIS key-values:
- **zombie-e** => Timespan (in minutes) to evaluate a zombie state (default = 5 min)
- **zombie-mdc** =>  Maximum distance (in meters) that a zombie can cover during zombie-e timespan (default = 500 m)

### Bonus point 2
The driver-location service accepts an optional  `distance=true` querystring option to recover the elapsedDistance and cumulativeDistance in a given timespan.

Example:

**Request**

`GET /drivers/:id/locations?distance=true&minutes=5`

**Response**
```
[
    {
        "cumulativeDistance": 0,
        "elapsedDistance": 0,
        "latitude": 48.864193,
        "longitude": 2.364986,
        "updated_at": "2018-10-24T13:58:10Z"
    },
    {
        "cumulativeDistance": 73.4,
        "elapsedDistance": 73.4,
        "latitude": 48.864193,
        "longitude": 2.365989,
        "updated_at": "2018-10-24T13:58:24Z"
    },
    {
        "cumulativeDistance": 146.407,
        "elapsedDistance": 73.007,
        "latitude": 48.864193,
        "longitude": 2.366987,
        "updated_at": "2018-10-24T14:00:17Z"
    }
]
```

CHANGELOG.md
## 1.0.0 (Oct 25, 2018)

Initial release
  - Gateway (.\gateway)
  - Driver-location service
  - Zombie-driver service
  - Tests

gateway/gateway_test.go
/*
Gateway service for Zombie test.

*/

package main

import (
	"bytes"
	"encoding/json"
	"log"
	"net/http"
	"net/http/httptest"
	"reflect"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestIniConfig_getConfFromYaml(t *testing.T) {
	type args struct {
		fileName string
	}
	var config IniConfig
	var expectedConfig IniConfig
	expectedConfig.Urls = []Endpoints{
		{"/drivers/:id/locations", "PATCH", NsqServiceOptions{"locations", "localhost:4151"}, HTTPRestServiceOptions{}},
		{"/drivers/:id", "GET", NsqServiceOptions{}, HTTPRestServiceOptions{"zombie-driver"}},
	}
	expectedConfig.Port = 3000
	var ar args
	ar.fileName = "../test/test-config.yaml"
	tests := []struct {
		name       string
		conf       IniConfig
		args       args
		wantResult IniConfig
		wantErr    bool
	}{
		// Test cases.
		{"Normal parse", config, ar, expectedConfig, false},
		{"Filename error", config, args{"../test/wrong.yml"}, config, true},
		{"File wrongly formatted", config, args{"../test/test-wrong-formatted.yaml"}, config, true},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			gotResult, err := tt.conf.getConfFromYaml(tt.args.fileName)
			if err != nil {
				t.Logf("Error detected   #%v ", err)
			}
			if (err != nil) != tt.wantErr {
				t.Errorf("IniConfig.getConfFromYaml() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !reflect.DeepEqual(gotResult, tt.wantResult) {
				t.Errorf("IniConfig.getConfFromYaml() = %v, want %v", gotResult, tt.wantResult)
			}
		})
	}
}

func TestNsqHandlerRoute(t *testing.T) {
	tests := []struct {
		name                string
		bodyAsString        string
		bodyPayload         map[string]interface{}
		expectedCode        int
		expectedBodyPayload interface{}
	}{
		//Test Cases
		{"Regular value", "", map[string]interface{}{"longitude": 2.364988, "latitude": 48.864193}, http.StatusOK, "Got data!"},
		{"Long is a string", "", map[string]interface{}{"longitude": "2.364988", "latitude": 48.864193}, http.StatusBadRequest, "Longitude is not a number"},
		{"Lat is a string", "", map[string]interface{}{"longitude": 2.364988, "latitude": "48.864193"}, http.StatusBadRequest, "Latitude is not a number"},
		{"Missing Long", "", map[string]interface{}{"latitude": 48.864193}, http.StatusBadRequest, "Missing latitude/longitude"},
		{"Missing Lat", "", map[string]interface{}{"longitude": 48.864193}, http.StatusBadRequest, "Missing latitude/longitude"},
		{"Empty body", "", map[string]interface{}{}, http.StatusBadRequest, "Missing latitude/longitude"},
		{"Not a JSON", "ahahaha", map[string]interface{}{}, http.StatusBadRequest, "Body is not a JSON."},
	}
	for _, tt := range tests {
		log.Printf("Testing %v", tt.name)
		router := setupRouter()
		body, _ := json.Marshal(tt.bodyPayload)
		if tt.bodyAsString != "" {
			body = []byte(tt.bodyAsString)
		}
		w := httptest.NewRecorder()
		req, _ := http.NewRequest("PATCH", "/drivers/test001/locations", bytes.NewBuffer(body))
		req.Header.Set("Content-type", "application/json")
		router.ServeHTTP(w, req)
		assert.Equal(t, tt.expectedCode, w.Code, "Testing "+tt.name)
		assert.Equal(t, tt.expectedBodyPayload, w.Body.String(), "Testing "+tt.name)
	}
}

func TestHttpForwardRoute(t *testing.T) {
	router := setupRouter()
	w := httptest.NewRecorder()
	req, _ := http.NewRequest("GET", "/drivers/test001", nil)
	router.ServeHTTP(w, req)

	assert.Equal(t, 200, w.Code)
	assert.Equal(t, "{\n    \"id\": \"test001\",\n    \"zombie\": true\n}", w.Body.String())
}


driver-location/driver-location.go
/*
Driver location service for Zombie test.

*/
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"math"
	"net/http"
	"reflect"
	"strconv"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/gomodule/redigo/redis"
	nsq "github.com/nsqio/go-nsq"
	yaml "gopkg.in/yaml.v2"
)

//IniConfig describes the data structure found config.yml file
type IniConfig struct {
	Port  int                 `yaml:"port,omitempty"`  //Gateway listening port
	Redis RedisServiceOptions `yaml:"redis,omitempty"` //Redis options
	Nsq   NsqServiceOptions   `yaml:"nsq,omitempty"`   //Nsq options
}

//RedisServiceOptions describes the options for Redis service
type RedisServiceOptions struct {
	Host     string `yaml:"host,omitempty"`     //host:port to connect Redis clients
	Password string `yaml:"password,omitempty"` //Password for AUTH command
}

//NsqServiceOptions describes the options for the driver-location service to interact with NSQ messaging service
type NsqServiceOptions struct {
	NsqlookupdHost string `yaml:"nsqlookupd-host,omitempty"` //nsqlookupd host:port that listens to native clients
	Topic          string `yaml:"topic,omitempty"`           //Topic to look for messages
	ChannelName    string `yaml:"channel,omitempty"`         //Channel name assigned to service's consumer
	MaxInflight    int    `yaml:"max-inflight,omitempty"`    //Maximum number of messages to allow in flight (concurrency knob)
	NumPublishers  int    `yaml:"num-publishers,omitempty"`  //number of concurrent publishers
}

//GLOBAL CONSTANTS

//ConfigFileName Path of the config file
const ConfigFileName string = "./config.yaml"

//DefaultMins Default minutes given by getLocations
const DefaultMins float64 = 5

//ChannelName Default NSQ channel name
const ChannelName = "driver-location-service"

//MaxReturnedElements Specifies how many elements can be returned in a Redis SORT request
//DEPRECATED - Evaluated dynamically according to the requested time length
//const MaxReturnedElements = 100

//GLOBAL VARIABLES

//Config is the struct that contains all the settings specified in config file
var Config IniConfig
var (
	pool *redis.Pool    //redis connection pool
	wg   sync.WaitGroup //Global waitgroup
)

func init() {
	//Loads config values from YAML file and sets them to Config global variable
	yamlConfig, err := Config.getConfFromYaml(ConfigFileName)
	if err != nil {
		log.Fatalf("Driver-location service can't be initialized. Exiting  %v", "")
	}
	log.Printf("Config values: %+v\n", yamlConfig)
	//Updates Config global variable
	Config = yamlConfig
	//TODO: Checks for minimal informations in config file and provide default values

}

//getConfFromYaml Extracts the config values from YAML config file
func (conf IniConfig) getConfFromYaml(fileName string) (result IniConfig, err error) {
	yamlFile, err := ioutil.ReadFile(fileName)
	if err != nil {
		log.Printf("yamlFile.Get err   #%v ", err)
		return conf, err
	}
	err = yaml.Unmarshal(yamlFile, &conf)
	if err != nil {
		log.Printf("Unmarshal: %v", err)
		return conf, err
	}
	return conf, nil
}

func timestampAsISO(ts int64) (ISOstring string) {
	t := time.Unix(ts, 0)
	JavascriptISOString := "2006-01-02T15:04:05Z07:00" //same as time.RFC3339
	ISOstring = t.UTC().Format(JavascriptISOString)
	return ISOstring
}

//newPool Creates a new Redis connection pool
func newPool(addr string) *redis.Pool {
	p := &redis.Pool{
		MaxIdle:     3,
		IdleTimeout: 240 * time.Second,
		Dial: func() (redis.Conn, error) {
			/*c, err := redis.Dial("tcp", addr)
			if err != nil {
				log.Printf("Error in connecting to Redis: %v", err)
				return nil, err
			}
			log.Printf("Connected!")
			return c, nil
			*/
			return redis.Dial("tcp", addr)
		},
	}
	return p
}

//getLocations Replies with an array of driver locations in the requested timespan
func getLocations(c *gin.Context) {
	//Reads minutes from querystring
	var minutes string
	minutes = c.DefaultQuery("minutes", "5")
	min, err := strconv.ParseFloat(minutes, 64)
	if err != nil {
		//Sets the default values in case of erroneous input
		min = DefaultMins
	}
	//Reads the distance flag from the querystring
	var distanceReq string
	wantsDistance := false //By default the distance computation is not required
	distanceReq = c.DefaultQuery("distance", "false")
	if distanceReq == "true" {
		wantsDistance = true
	}
	//Reads driverId from the path params
	id := c.Param("id")
	//Retrieves the eligible timestamps info from REDIS
	//Evaluate Now() timestamp (Unix time)
	now := time.Now().Unix()
	//Gets a list of recorded timestamps for driver:id
	if pool == nil {
		//Pool not initialized (e.g. in unit tests). Provide a pool initialization
		log.Println("Redis pool not initialized. Proceed with initialization")
		pool = newPool(Config.Redis.Host)
	}
	conn := pool.Get()
	defer conn.Close()
	reply, err := redis.Int64s(conn.Do("SORT", fmt.Sprintf("driver:%v:timestamps", id), "LIMIT", 0, min*60, "DESC"))
	if err != nil {
		log.Printf("Error in processing SORT request. %v", err)
		c.String(http.StatusInternalServerError, "Ooops. Something went wrong on our side.")
		return
	}
	log.Printf("Got timestamp list. %v", reply)
	//Empty timestamp list? --> driver doesn't exist. Reply with 404 error
	if len(reply) == 0 {
		notFoundReply := map[string]string{
			"message": "Driver not found",
		}
		c.IndentedJSON(http.StatusNotFound, notFoundReply)
		return
	}
	eligibleTimestamps := make([]int64, 0)
	for i := 0; i < len(reply); i++ {
		timestamp := reply[i]
		if float64(timestamp) >= float64(now)-(min*60) {
			eligibleTimestamps = append(eligibleTimestamps, reply[i])
		} else {
			//Sorted array. There are no more interesting timestamps. Ends for cycle
			i = len(reply)
		}
	}
	//Builds the response
	response := make([]map[string]interface{}, 0)
	var total float64 //total Holds the total distance that a driver made during "minutes"
	total = 0
	for i := 0; i < len(eligibleTimestamps); i++ {
		//Since eligibleTimestamps is ordered by DESC and the response is ordered by ASC we use a decreasing cursor
		j := len(eligibleTimestamps) - 1 - i
		timestamp := eligibleTimestamps[j]
		reply, err := redis.Positions(conn.Do("GEOPOS", fmt.Sprintf("driver:%v:log", id), timestamp))
		if err != nil {
			//Error in executing GEOPOS. Go on with next i
			log.Printf("Error in processing GEOPOS request. %v", err)
			log.Println("Going on to retrieve GEOPOS for other eligible timestamps")
		} else {
			//Reply contains long and lat. Add them to response and round to 6 digits precision
			newElement := make(map[string]interface{})
			newElement["latitude"] = math.Floor(reply[0][1]*1e6) / 1e6
			newElement["longitude"] = math.Floor(reply[0][0]*1e6) / 1e6
			newElement["updated_at"] = timestampAsISO(timestamp)
			//Check if it has to add distance and delta in the response
			if wantsDistance {
				//Evaluates the distance between eligibleTimestamps[j] and eligibleTimestamps[j-1]
				var delta float64
				if i == 0 {
					//The first element is the start for evaluating deltas
					delta = 0
				} else {
					//Retrieves delta
					delta, err = redis.Float64(conn.Do("GEODIST", fmt.Sprintf("driver:%v:log", id), eligibleTimestamps[j], eligibleTimestamps[j+1], "m"))
				}
				//Updates total (if GEODIST calls gives an error, delta = 0)
				total = total + delta
				//Adds delta and total to the response
				newElement["elapsedDistance"] = math.Floor(delta*1e3) / 1e3
				newElement["cumulativeDistance"] = math.Floor(total*1e3) / 1e3
			}
			//Updates response slice
			response = append(response, newElement)
		}
	}
	//Sends the response
	c.IndentedJSON(http.StatusOK, response)
	return
}

func validateInput(input map[string]interface{}) bool {
	//At the beginning the return value isValidated is set to "true"
	isValidated := true
	//1. Are there the necessary fields?
	longitude, isThereLongitude := input["longitude"]
	latitude, isThereLatitude := input["latitude"]
	id, isThereID := input["driverId"]
	if !isThereLongitude || !isThereLatitude || !isThereID {
		//Missing fields
		log.Println("Missing fields")
		isValidated = false
		//Returns immediately to avoid missing fields parsing attempts
		return isValidated
	}
	//2. Checks if the fields are ther with nil values
	if id == nil || latitude == nil || longitude == nil {
		log.Println("Some nil values")
		isValidated = false
		//Returns immediately to avoid nil fields parsing attempts
		return isValidated
	}
	//3. Fields are there. Check if they are correctly typed
	//long/lat
	if reflect.TypeOf(longitude).String() != "float64" || reflect.TypeOf(latitude).String() != "float64" {
		log.Println("Wrong numeric type")
		log.Printf("long: %v, lat: %v", reflect.TypeOf(longitude).String(), reflect.TypeOf(latitude).String())
		isValidated = false
		return isValidated
	}
	//id
	if !(reflect.TypeOf(id).String() == "string" || reflect.TypeOf(id).String() == "float64" || reflect.TypeOf(id).String() == "int") {
		//for our use, id could be a string or a number  -> json.unmarshall parse numbers as float64. Int is kept for future possibilities
		log.Println("Wrong id type")
		isValidated = false
		return isValidated
	}
	//4. Check the emptyness/in range values only if the validation is true so far
	if isValidated {
		//Checks if the fields are not empty or have invalid values
		if longitude.(float64) > 180 || longitude.(float64) < -180 || latitude.(float64) > 85.05112878 || latitude.(float64) < -85.05112878 || id == "" {
			log.Println("Invalid value")
			isValidated = false
			return isValidated
		}
	}
	return isValidated
}

//persistMessageToRedis Saves a valid message to an appropriate set of key-values in Redis
func persistMessageToRedis(message map[string]interface{}) error {
	//Gets a connection from the Pool
	if pool == nil {
		//Pool not initialized (e.g. in tests). Create a new pool
		pool = newPool(Config.Redis.Host)
	}
	conn := pool.Get()
	defer conn.Close()
	//Checks if there have been problems in connecting
	if conn.Err() != nil {
		log.Printf("Error in connecting to Redis: %v", conn.Err())
		return conn.Err()
	}
	//Saves the instant position
	writeErrors := make([]error, 0)
	timestamp := message["timestamp"].(int64) //timestamp in Unix time
	latitude := message["latitude"].(float64)
	longitude := message["longitude"].(float64)
	id := message["driverId"]
	//On-course key. Adds the position of driver id
	_, err := conn.Do("GEOADD", "on-course", longitude, latitude, id)
	if err != nil {
		writeErrors = append(writeErrors, err)
		log.Printf("Error in saving on-course data with GEOADD: %v", err)
	}
	//driver:id:log key. Adds postion & timestamp
	_, err = conn.Do("GEOADD", fmt.Sprintf("driver:%v:log", id), longitude, latitude, timestamp)
	if err != nil {
		writeErrors = append(writeErrors, err)
		log.Printf("Error in saving driver log data with GEOADD: %v", err)
	}
	//dirver:id:timestamps. Adds the recorded timestamp to a list connected to driver:id
	_, err = conn.Do("SADD", fmt.Sprintf("driver:%v:timestamps", id), timestamp)
	if err != nil {
		writeErrors = append(writeErrors, err)
		log.Printf("Error in saving driver timestamp data with SADD: %v", err)
	}
	//Check if there have been errors in redis writes
	if len(writeErrors) > 0 {
		finalError := fmt.Errorf("There have been errors in redis writes: %v", writeErrors)
		return finalError
	}
	//Exits the method with no errors
	log.Println("Message have been persisted successfully to Redis")
	return nil
}

//handleMessage Handles what to do when a message from NSQ topic/channel is received
func handleMessage(m *nsq.Message) error {
	//log.Printf("Message received: %+v", *m)
	log.Printf("Message body: %v", string(m.Body))
	//Extracts the timestamp in Unix format
	timestamp := m.Timestamp / 1e9
	//JSON is already validated by downstream service (Gateway). Unmarshal it in a generic map
	var parsedMessage map[string]interface{}
	err := json.Unmarshal(m.Body, &parsedMessage)
	if err != nil {
		//Wrong JSON decoding. Return nil to avoid requeuing.
		log.Printf("Something went wrong while decoding the JSON message payload. %v", err)
		return nil
	}
	//Validate the input (message)
	if !validateInput(parsedMessage) {
		//Message hasn't valid format. Return nil to avoid requeuing.
		log.Println("Message has not a valid structure and won't be persisted")
		return nil
	}
	//Input is validated. Add timestamp and send it to Redis
	parsedMessage["timestamp"] = timestamp
	err = persistMessageToRedis(parsedMessage)
	if err != nil {
		//Logs the error but doesn't return an error to the handler (fails silently and avoid requeing)
		log.Printf("An error occured while calling persistMessageToRedis: %v", err)
	}
	return nil
}

//poolNSQForMessages Pools messages from NSQ service
func poolNSQForMessages() {
	cfg := nsq.NewConfig()
	cfg.DialTimeout = 10 * time.Second
	cfg.UserAgent = fmt.Sprintf("driver-location/%s go-nsq/%s", "0.1", nsq.VERSION)
	cfg.MaxInFlight = Config.Nsq.MaxInflight
	consumer, err := nsq.NewConsumer(Config.Nsq.Topic, Config.Nsq.ChannelName, cfg)
	if err != nil {
		log.Fatalf("A problem occurred in initializing NSQ Consumer: %v", err)
		return
	}
	consumer.AddConcurrentHandlers(nsq.HandlerFunc(handleMessage), Config.Nsq.NumPublishers)
	nsqlookupAddresses := make([]string, 0)
	nsqlookupAddresses = append(nsqlookupAddresses, Config.Nsq.NsqlookupdHost)
	err = consumer.ConnectToNSQLookupds(nsqlookupAddresses)
	if err != nil {
		log.Printf("A problem occured in connecting to nsqlookupd: %v", err)
		return
	}
	log.Println("Connected to nsdlookupd")
}

//setupRouter Defines the routes exposed by driver-location service
func setupRouter() *gin.Engine {
	router := gin.Default()
	router.GET("/drivers/:id/locations", getLocations)
	return router
}

// routing Sets up the Gin framework router
func routing() {
	//Sets the routes
	router := setupRouter()
	//Starts the gateway
	if Config.Port == 0 {
		router.Run()
	} else {
		router.Run(":" + strconv.Itoa(Config.Port))
	}
}

//Main routine
func main() {
	//Creates a Redis pool and sets it to a module wide variable
	pool = newPool(Config.Redis.Host)
	log.Printf("Redis pool stats: %v", pool.Stats())
	//Starts to pool NSQ for location messages
	poolNSQForMessages()
	//Sets up the Gin framework router in a separate goroutine
	wg.Add(1) //Adds 1 to waitgroup
	go routing()
	log.Println("End of main() calls")
	//Avoids that main() ends with goroutines still in execution
	wg.Wait()
}
